### PR TITLE
Add support of Date32 arguments for date_diff function

### LINKED ISSUE #41977: Add support of Date32 arguments for date_diff function
**Use case**

Currently date_diff function accepts only Date, DateTime and DateTime64 arguments but it does not support Date32 arguments. 

For example:

`select date_diff('day', toDate32('1979-01-01'), toDate32('1980-01-01'))`

Returns:

> DB::Exception: Second argument for function dateDiff must be Date or DateTime

**Describe the solution you'd like**

Implement support of Date32 arguments.

### PR DESCRIPTION
**Pull request by rvasin**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add a possibility to use Date32 arguments for date_diff function. Fix issue in date_diff function when using DateTime64 arguments with start date before Unix epoch and end date after Unix epoch.

Close #41977 and #41978.

### CODE REVIEW

**Comment by rvasin** [comment]
@rschu1ze Please be reviewer of this PR and assign "can be tested" label.

**Comment by rschu1ze** [comment]
Sure, thanks. (Please expect a bit of delay for the review, I am in transit right now and need to catch up with some stuff.)

**Comment by rvasin** [comment]
OK. It's not a problem. It's not urgent. I am planning two more PRs to make the remaining functions to support Date32 arguments.

**Comment by rvasin** [comment]
@alexey-milovidov could you please give your expert opinion on dateDiff function? 
dateDiff uses toRelative function in this way:

`dateDiff(unit, A, B) = toRelative(B) - toRelative(A) `

@ilejn noticed that this rule maybe be considered as a some kind of constraint. Thus maybe some ClickHouse users rely on this constraint: they may think that dateDiff is the difference of relatives dates.

In order to implement #41977  and fix #41978 we need to modify toRelative() because it overflows on negative values (and there are several other aspects see toStableRelativeHourNum). As I discussed with @rschu1ze we decided to keep the existing toRelative() function as it is. Because these functions maybe used in many other places and we decided to implement internal variants of toRelative functions (toExtendedRelative) which return signed results. 

@ilejn said that this approach may break the constraint. In my opinion there is no constraint at all. Someone selected to use toRelative() functions to easily implement dateDiff. But dateDiff() maybe implemented even without using of toRelative(). In my opinion it’s important that dateDiff()  works correctly (it returns correct results for given arguments) and it’s does not matter how it’s implemented inside. 

@alexey-milovidov please let us know does the current approach in this PR breaks the constraint? And does it even matter to keep the constraint?

File: src/Common/DateLUTImpl.h [hunk 1]
```diff
@@ -895,6 +895,19 @@ class DateLUTImpl
         return toRelativeHourNum(lut[toLUTIndex(v)].date);
     }
 
+    /// The same formula is used for positive time (after Unix epoch) and negative time (before Unix epoch).
```
Comment (rschu1ze) [inline src/Common/DateLUTImpl.h:898]: Why exactly can `toRelativeHourNum` not be used from "date_diff"?

File: tests/queries/0_stateless/02458_datediff_date32.reference [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+1
+SELECT dateDiff('week', toDate('2015-08-18'), toDate32('2015-08-25'));
+1
+SELECT dateDiff('month', toDate('2015-08-18'), toDate32('2015-09-18'));
+1
+SELECT dateDiff('quarter', toDate('2015-08-18'), toDate32('2015-11-18'));
+1
+SELECT dateDiff('year', toDate('2015-08-18'), toDate32('2016-08-18'));
+1
+-- Const vs non-const columns
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02458_datediff_date32.reference:125]: With the internal implementation in mind, the const vs. non-const are a bit overkill and could be omitted.

**Comment by rschu1ze** [comment]
In my POV, the only constraint users may righteously claim to rely on is what the [documentation](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#date_diff) specifies:

```
Returns the difference between two dates or dates with time values. The difference is calculated using relative units, 
e.g. the difference between 2022-01-01 and 2021-12-29 is 3 days for day unit (see toRelativeDayNum), 1 month for 
month unit (see toRelativeMonthNum), 1 year for year unit (see toRelativeYearNum).
```

With reference to [my comment](https://github.com/ClickHouse/ClickHouse/issues/41977#issuecomment-1267536814), we are in "a)-territory" (i.e. providing new functionality doesn't break backward compact) BUT the documentation actually also constraints the new functionality.

The documentation actually corresponds to formula `dateDiff(unit, A, B) = toRelative(B) - toRelative(A)`. So I would agree with @ilejn.

Question: Under which circumstances does this actually matter? As far as I understand, this PR follows the constraint already. Did I miss something?

File: tests/queries/0_stateless/02458_datediff_date32.reference (hunk 2 above)
Comment (rvasin) [reply tests/queries/0_stateless/02458_datediff_date32.reference:125]: As a base for writing 2458_datediff_date32 test I took tests/queries/0_stateless/01801_dateDiff_DateTime64.sql test which also has "const vs. non-const" blocks. BTW after finishing my 2458_datediff_date32 test and running it I found the forgotten code in "const vs. non-const" operations (I created  commit [14ba9fa](https://github.com/ClickHouse/ClickHouse/pull/42308/commits/14ba9fa51da3d505c3e040460600b07687400432) for it). So I think keeping the "const vs. non-const"  blocks of test make sense.

File: src/Functions/dateDiff.cpp [hunk 3]
```diff
@@ -33,6 +33,199 @@ namespace ErrorCodes
 namespace
 {
 
+struct ToExtendedRelativeYearNumImpl
```
[the remark this answers is not in the document]
Comment (rvasin) [reply src/Functions/dateDiff.cpp:36]: I agree this approach introduces less "code bloating". I probably like it more than current approach. While it makes extended variants of toRelative() to be some kind of public functions (while there are not strictly public because they are protected by ExtendedPrecision), while in my current code they are private - they don't go outside of dateDiff.cpp.

So I will rewrite my code to use your approach. 

BTW what about naming of the ExtendedPrecision? It's not about precision but it's about extended result. Maybe ExtendedResult is a better name? 

Also maybe make it `false` by default? 
` template<bool ExtendedPrecision = false>`

[the remark this answers is not in the document]
Comment (rvasin) [reply src/Functions/dateDiff.cpp:36]: Also in previous PR I implemented:
src/Functions/DateTimeTransforms.h

```
template <typename FromDataType, typename ToDataType, typename Transform, bool is_extended_result = false>
struct DateTimeTransformImpl
```

In CH guides the variable naming should be in extended_precision and not ExtendedPrecision. 

So maybe simply use is_extended_result  for toRelative functions.

File: src/Functions/dateDiff.cpp [hunk 4]
```diff
@@ -61,25 +254,30 @@ class FunctionDateDiff : public IFunction
     DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
     {
         if (arguments.size() != 3 && arguments.size() != 4)
-            throw Exception("Number of arguments for function " + getName() + " doesn't match: passed "
-                + toString(arguments.size()) + ", should be 3 or 4",
-                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 3 or 4",
+                getName(), toString(arguments.size()));
```
[the remark this answers is not in the document]
Comment (rvasin) [reply src/Functions/dateDiff.cpp:259]: Done. Thank you for noticing it.

File: src/Functions/dateDiff.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rvasin) [reply src/Functions/dateDiff.cpp]: Implemented it the way described above and pushed. 
While use it this way:
```
template<bool is_extended_result>
struct ToRelativeMonthNumImpl
```
So use it without default value. Strange, with default value:
```
template<bool is_extended_result = false>
struct ToRelativeMonthNumImpl
```

It does not compile:

> toRelativeMonthNum.cpp:10:86: error: use of class template 'ToRelativeMonthNumImpl' requires template arguments
> [build] using FunctionToRelativeMonthNum = FunctionDateOrDateTimeToSomething<DataTypeUInt32, ToRelativeMonthNumImpl>;
> 


In ToRelativeHourNumImpl please see:

```
    static inline UInt32 execute(UInt32 t, const DateLUTImpl & time_zone)
    {
        if constexpr (is_extended_result)
            return time_zone.toStableRelativeHourNum(static_cast<time_t>(t));
        else
            return time_zone.toRelativeHourNum(static_cast<time_t>(t));
    }
    // ... skip ...
    static inline UInt32 execute(UInt16 d, const DateLUTImpl & time_zone)
    {
        if constexpr (is_extended_result)
            return time_zone.toStableRelativeHourNum(DayNum(d));
        else
            return time_zone.toRelativeHourNum(DayNum(d));
    }
```
So usage of ``is_extended_result`` is not strictly extended. Because the result type is the same, but in this case it's rather a compatibility handling. 
It's needed to do calculations "via" Unix epoch. I will explained in in another thread of this PR.

File: src/Common/DateLUTImpl.h (hunk 1 above)
Comment (rvasin) [reply src/Common/DateLUTImpl.h:898]: @rschu1ze Consider the function:

```
inline Time toRelativeHourNum(Time t) const
{
    if (t >= 0 && offset_is_whole_number_of_hours_during_epoch)
        return t / 3600;

    /// Assume that if offset was fractional, then the fraction is the same as at the beginning of epoch.
    /// NOTE This assumption is false for "Pacific/Pitcairn" and "Pacific/Kiritimati" time zones.
    return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
}

```

For positive values (dates after Unix epoch) we use the algorithm:

```
if (t >= 0 && offset_is_whole_number_of_hours_during_epoch)
    return t / 3600;

```    

For negative values (dates before Unix epoch) we use the algorithm:

`return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
`
This way it works fine with both arguments before or both after Unix epoch.

But if start date is before Unix epoch and end date is after Unix epoch then it produces a bias mainly because of using 86400 - offset_at_start_of_epoch for dates before Unix epoch. So it does not work, see #41978

That why in toStableRelativeHourNum we need to use the same formula:

`return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
`
to calculate relative dates before and after Unix epoch. 

Consider the function:

```
inline Time toRelativeMinuteNum(Time t) const /// NOLINT
{
    return (t + DATE_LUT_ADD) / 60 - (DATE_LUT_ADD / 60);
}

```    
So it uses the same formula for negative and positive dates.
    
But the "big" question is why toRelativeHourNum is "time-zone aware" while toRelativeMinuteNum does not know about time zones? Why what we do for **hours** we don't do for **minutes**?

File: src/Common/DateLUTImpl.h [hunk 5]
```diff
@@ -895,6 +895,19 @@ class DateLUTImpl
...
+    /// It’s needed for correct work of dateDiff function.
+    inline Time toStableRelativeHourNum(Time t) const
```
[the remark this answers is not in the document]
Comment (rvasin) [reply src/Common/DateLUTImpl.h:900]: Done. Added `toStableRelativeHourNum` to src/Common/tests/gtest_DateLUTImpl.cpp

File: src/Common/DateLUTImpl.h (hunk 1 above)
Comment (rvasin) [reply src/Common/DateLUTImpl.h:898]: I also don't understand the meaning of the calculation:

`return (t + DATE_LUT_ADD) / 60 - (DATE_LUT_ADD / 60);`

Because the result is about the same as:

`return t / 60;`

If we want to make result type non-negative then we need to do just:
    
`return (t + DATE_LUT_ADD) / 60;`

But `toRelativeMinuteNum` returns `Time` which is signed `Int64`. So it's OK to have negative times before Unix epoch.

Comment (rvasin) [reply src/Common/DateLUTImpl.h:898]: BTW I found the following issues which are reproducted in current master (should we report it as a separate issue? or probably fix it in current PR?)

It's related when start date is Unix epoch:
```

SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-02 00:00:00'));
-4294965856 - wrong, must be 1440

SELECT dateDiff('minute', toDateTime('1970-01-02 00:00:00'), toDate('1970-01-01'));
4294965856 - wrong, must be -1440

SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-03 00:00:00'));
-4294964416 - wrong, must be 2880

SELECT dateDiff('minute', toDate('1970-01-02'), toDateTime('1970-01-03 00:00:00'));
1440 - OK

SELECT dateDiff('minute', toDate('1970-01-03'), toDateTime('1970-01-04 00:00:00'));
1440 - OK
```

File: src/Functions/dateDiff.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/dateDiff.cpp]: > BTW what about naming of the ExtendedPrecision? It's not about precision but it's about extended result. Maybe ExtendedResult is a better name?

Agree, `ExtendedResult` sounds better. Didn't put much thought into naming.

> Also maybe make it false by default?
> `template<bool ExtendedPrecision = false>`

I have witnessed too many unnecessary disasters with default function parameters or default template parameters. See e.g. #37117 for a recent example in ClickHouse which led to wrong results. My personal preference would be to make stuff explicit, unless it is a ton of work.

File: src/Common/DateLUTImpl.h (hunk 1 above)
Comment (rvasin) [reply src/Common/DateLUTImpl.h:898]: Probably we need to create new function:
toStableRelativeMinuteNum
and make it time zone "aware":

```
inline Time toStableRelativeMinuteNum(Time t) const /// NOLINT
{
    return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 60 - (DATE_LUT_ADD / 60);
}
```
So it will be used in dateDiff only, the same way as toStableRelativeHourNum now. 
And probably it will fix the issue above.

File: src/Functions/dateDiff.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/dateDiff.cpp]: Totally optional, but if you like to make this even more fancy, you could use a scoped enum as template parameter

```cpp
enum class ResultPrecision
{
    Standard,
    Extended
};

// in the template:
template <ResultPrecision precision_>
class ToRelativeYearNumImpl

   if constexpr(precision_ == ResultPrecision::Standard)
   {
        ...
   }

// the main advantage is that the template instantiation becomes prettier:

ToRelativeYearNumImpl<ResultPrecision::Standard> impl;
```

File: src/Common/DateLUTImpl.h (hunk 1 above)
Comment (rvasin) [reply src/Common/DateLUTImpl.h:898]: BTW the issue is re-produced if non-UTC timezone is specified in config.xml, for example:
<timezone>Europe/Paris</timezone>

But if time-zone is not defined or it's defined as UTC
<timezone>UTC</timezone>

Then it works OK:
```
SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-02 00:00:00'));
1440
```
Maybe it's because Date is not time-zone aware and DateTime is timezone aware.

File: src/Functions/dateDiff.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/dateDiff.cpp]: > It does not compile:

>     toRelativeMonthNum.cpp:10:86: error: use of class template 'ToRelativeMonthNumImpl' requires template arguments
>    [build] using FunctionToRelativeMonthNum = FunctionDateOrDateTimeToSomething<DataTypeUInt32, ToRelativeMonthNumImpl>;

I think the syntax to use a default template parameter would be `ToRelativeMonthNumImpl<>`.

[the remark this answers is not in the document]
Comment (rvasin) [reply src/Functions/dateDiff.cpp]: I re-wrote the code to use ResultPrecision::Standard. Please take a look. Yes, it's prettier and it's more easy to read / understand such code.

**Review by rschu1ze — APPROVED** [review]

File: tests/queries/0_stateless/02457_datediff_via_unix_epoch.sql [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+select 'day', date_diff('day', toDateTime64('1969-12-25 10:00:00.000', 3), toDateTime64('1970-01-05 10:00:00.000', 3));
+
+select 'hour', date_diff('hour', toDate32('1969-12-25'), toDate32('1970-01-05'));
+select 'hour', date_diff('hour', toDateTime64('1969-12-25 10:00:00.000', 3), toDateTime64('1970-01-05 10:00:00.000', 3));
+
+select 'minute', date_diff('minute', toDate32('1969-12-31'), toDate32('1970-01-01'));
+select 'minute', date_diff('minute', toDateTime64('1969-12-31 23:50:00.000', 3), toDateTime64('1970-01-01 00:10:00.000', 3));
+
+select 'second', date_diff('second', toDate32('1969-12-31'), toDate32('1970-01-01'));
+select 'second', date_diff('second', toDateTime64('1969-12-31 23:50:00.000', 3), toDateTime64('1970-01-01 00:10:00.000', 3));
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02457_datediff_via_unix_epoch.sql:23]: Since `toStableRelativeHourNum` is called for possible argument types (not only `Date32`), it would surely be good to add tests for Date/DateTime/DateTime64 arguments < and > 1970 as well. But totally fine to do this in a separate PR.

File: src/Common/DateLUTImpl.h (hunk 1 above)
Comment (rschu1ze) [reply src/Common/DateLUTImpl.h:898]: I read the earlier discussion (https://github.com/ClickHouse/ClickHouse/pull/42308#issuecomment-1282662137, https://github.com/ClickHouse/ClickHouse/pull/42308#issuecomment-1282749612). I think I misunderstood what was meant by "constraint".

The [documentation](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#date_diff) describes that `date_diff` has the intuitive behavior, i.e. how many sec/min/h/d/w/m/q/y-s do I need to add or subtract to get from one date/time to the other. The function just happens to be implemented using `toRelativeUnitNum()` because that is convenient from a programming POV. More specifically, there is *no* obligation/constraint to replicate (overflow) bugs in `toRelativeUnitNum()` into `date_diff()`, in fact I would consider bug-by-bug-compatibility a bug in itself 😄

Comment (rschu1ze) [reply src/Common/DateLUTImpl.h:898]: My understanding is that this PR technically breaks backward compact but only in a quite obscure edge case, namely for `date_diff('hour', A, B)` where A and B are 'DateTime64' and smaller and bigger than 1970-01-01 respectively. Besides that, it adds support for `Date32` arguments to `date_diff` but that cannot be considered backward incompatible.

File: tests/queries/0_stateless/02458_datediff_date32.reference [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+SELECT dateDiff('day', toDate32('1900-01-01'), toDateTime64('1900-01-02 00:00:00', 3));
+1
+SELECT dateDiff('week', toDate32('1900-01-01'), toDateTime64('1900-01-08 00:00:00', 3));
+1
+SELECT dateDiff('month', toDate32('1900-01-01'), toDateTime64('1900-02-01 00:00:00', 3));
+1
+SELECT dateDiff('quarter', toDate32('1900-01-01'), toDateTime64('1900-04-01 00:00:00', 3));
+1
+SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));
+1
```
Comment (azat) [inline tests/queries/0_stateless/02458_datediff_date32.reference:37]: I'm not sure is this change breaks it or not, likely it is introduced in #39425, but I will write here since anyway you are the author of both and it seems that you really use this extended support.

This line is flaky - https://play.clickhouse.com/play?user=play#c2VsZWN0IAogIHJlcG9ydF91cmwKZnJvbSBjaGVja3MKd2hlcmUKICBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gaW50ZXJ2YWwgNyBkYXkgYW5kCiAgdGVzdF9uYW1lIGxpa2UgJyUwMjQ1OF9kYXRlZGlmZl9kYXRlMzIlJyBhbmQKICB0ZXN0X3N0YXR1cyBpbiAoJ0ZBSUwnLCAnRkxBS1knLCAnU0tJUFBFRCcp

For `Pacific/Guam` timezone it has different result:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'UTC') AS utc,
    toDateTime64('1901-01-01 00:00:00', 3, 'Pacific/Guam') AS fail

Row 1:
──────
utc:  1901-01-01 00:00:00.000
fail: 1900-12-31 23:45:00.000
```

And this is for 22.7 (without referenced PR):

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'UTC') AS utc,
    toDateTime64('1901-01-01 00:00:00', 3, 'Pacific/Guam') AS fail

Row 1:
──────
utc:  1925-01-01 00:00:00.000
fail: 1925-01-01 00:00:00.000
```

Also I don't understand how DateTime64 equality works, is this a bug?

```sql
SELECT toDateTime('1901-01-01 00:00:00', 'UTC') = toDateTime('1901-01-01 00:00:00', 'Europe/Moscow') AS eq

┌─eq─┐
│  1 │
└────┘

SELECT toDateTime64('1901-01-01 00:00:00', 3, 'UTC') = toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow') AS eq

┌─eq─┐
│  0 │
└────┘
```

And one more thing, why `toDateTime64` differs from `toDateTime`:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow'),
    toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')

Row 1:
──────
toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow'): 1925-01-01 00:00:00.000
toDateTime('1901-01-01 00:00:00', 'Europe/Moscow'):      1970-01-01 03:00:00
``

Comment (rschu1ze) [reply tests/queries/0_stateless/02458_datediff_date32.reference:37]: So the fix would be to specify the time zone (e.g. `UTC`) in `toDateTime64()`. I have seen the same issue leading to flakiness in other tests too.

> Also I don't understand how DateTime64 equality works, is this a bug?

About
```sql
SELECT toDateTime('1901-01-01 00:00:00', 'UTC') = toDateTime('1901-01-01 00:00:00', 'Europe/Moscow') AS eq
```
Date type `DateTime` before 1970 is undefined, hence the comparison is meaningless.

`SELECT toDateTime64('1901-01-01 00:00:00', 3, 'UTC') = toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow') AS eq` gives the expected result (`0`).

> And one more thing, why toDateTime64 differs from toDateTime:

```sql
SELECT
    toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow')
    toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')
```

This gives on my system

```
┌─toDateTime64('1901-01-01 00:00:00', 3, 'Europe/Moscow')─┬─toDateTime('1901-01-01 00:00:00', 'Europe/Moscow')─┐
│                                 1901-01-01 00:00:00.000 │                                1970-01-01 03:00:00 │
└─────────────────────────────────────────────────────────┴────────────────────────────────────────────────────┘
```

Comment (rvasin) [reply tests/queries/0_stateless/02458_datediff_date32.reference:37]: @azat I made a series of PRs related to adding extended range support for Date/Time functions. But I did no modify toDateTime64 and toDateTime functions in my PRs.
I will take a look at it today.

As for toDateTime function - yes, as @rschu1ze said: the value is undefined before 1970-01-01 so any arguments for toDate or toDateTime function before 1970 like 1900 or 1901 will be "narrowed" to 1970-01-01.

**Comment by alexey-milovidov** [comment]
This PR contains a bug.
The function does not use the time zone if one of the arguments is Date.
Example:

```
$ clickhouse-local --query "SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));"
1
$ TZ=Pacific/Tarawa clickhouse-local --query "SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));"
0
```

I would expect, that the timezone `Pacific/Tarawa` will be applied for the interpretation of both arguments. For example, if `'1901-01-01 00:00:00'` is in Pacific/Tarawa, then '1900-01-01' should be also interpreted in this time zone, and the result will be always 1.

**Comment by rvasin** [comment]
@alexey-milovidov I will take a close look on it today and fix it if needed.
But please answer: is it true that Date and Date32 data types support time zones?
I know that DateTime and DateTime64 support time zones. But I am not sure about Date and Date32 data types.

**Comment by rvasin** [comment]
Also as I previously said: in my recent PRs related to date/time function I did not modify toDate/toDate32/toDateTime/toDateTime64 functions. So I don't know yet for sure if time time is correctly passed to toDate/toDate32 functions, for example.

**Comment by rvasin** [comment]
In my quick tests date_diff function works fine for different server's time zones:

```
timezone in config.xml: Europe/Moscow
SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));
It returns: 1

timezone in config.xml: Europe/Paris
SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));
It returns: 1

timezone in config.xml: UTC
SELECT dateDiff('year', toDate32('1900-01-01'), toDateTime64('1901-01-01 00:00:00', 3));
It returns: 1
```
I will try to re-run all test queries of @azat to give more information.

**Comment by rschu1ze** [comment]
I can observe the error only for `Date32`/`DateTime64` values very close to the lower border (1900-01-01) and with "eastern" timezones (e.g. "Pacific/Guam" or "Pacific/Tarawa"). It rather looks like a problem with `toDate32()`/ `toDateTime64()` than with `dateDiff()`.

**Comment by rvasin** [comment]
Robert, thank you for noticing it. As I said: the issue it probably related toDate/toDate32 function and is not related to date_diff itself. 

Also I remind that support of times zones in 1900..1925 is not reliable because of frequent time zones changes during that period and because time zones was not fractions of 15 minutes and so on. As @ alexey-milovidov  [said](https://github.com/ClickHouse/ClickHouse/issues/38393#issuecomment-1166176339):

> Extending DateLUT before 1925 with inaccurate support of time zones will be ok for me.

Date/time range was extended to 1900 because in banks the starting date is 1900 (it's considered as "zero"  date). So it was extended to 1900 mostly because of Date32 usage by banks (not because of frequent DateTime64 near 1900).

**Comment by rvasin** [comment]
> I would expect, that the timezone `Pacific/Tarawa` will be applied for the interpretation of both arguments. For example, if `'1901-01-01 00:00:00'` is in Pacific/Tarawa, then '1900-01-01' should be also interpreted in this time zone, and the result will be always 1.

This is an interesting idea if one of arguments is Date or Date32 and another argument is DateTime or DateTime64 than take the time zone from DateTime/DateTime64 argument and apply it to Date/Date32 argument. I will try to make some "experiments" with it.

**Comment by rschu1ze** [comment]
What is interesting though is that a query with two `Date32` arguments

```
TZ=Pacific/Tarawa ./clickhouse-local --query "SELECT dateDiff('year', toDate32('1900-01-01'), toDate32('1901-01-01 00:00:00'));"
```

works fine, i.e. result = 1 whereas with two `DateTime64` arguments

```
TZ=Pacific/Tarawa ./clickhouse-local --query "SELECT dateDiff('year', toDateTime64('1900-01-01', 3), toDateTime64('1901-01-01 00:00:00', 3));"
```

the result is = 0.

Since dateDiff forwards the computation to "toRelative*Num()", could there be a bug with "toRelative*Num()" with DateTime64 types?

**Comment by rvasin** [comment]
We we discussed earlier in this PR: `dateDiff` is using toRelative functions this way:

`dateDiff(unit, A, B) = toRelative(B) - toRelative(A) `

I localized the problem, it's because of toRelativeYearNum function:

```
$ TZ=Pacific/Guam ./clickhouse-local --query "SELECT toRelativeYearNum(toDateTime64('1900-01-01 00:00:00', 3));"
1900  - OK
$ TZ=Pacific/Guam ./clickhouse-local --query "SELECT toRelativeYearNum(toDateTime64('1901-01-01 00:00:00', 3));"
1900  - Wrong
$ TZ=Pacific/Guam ./clickhouse-local --query "SELECT toRelativeYearNum(toDateTime64('1902-01-01 00:00:00', 3));"
1902  - OK
```

So for TZ=Pacific/Guam and 1901 function `toRelativeYearNum` does not work well. It's not a problem of date_diff function. Important: it seems that toRelativeHourNum/toRelativeQuarterNum etc work fine for all time zones  in 1900, 1901.  It's only `toRelativeYearNum` which does not work well for just 1901.
    
@rschu1ze  Robert, I suggest to change year from 1900 to 1927 in tests/queries/0_stateless/02458_datediff_date32.sql 
It will eliminate flaky behavior of 02458_datediff_date32 test.
I may do it now and create a new PR for that.
What do you think?

**Comment by rschu1ze** [comment]
@rvasin Sounds good, let's do that. Thanks for checking.

Btw., the problem with toRelative*(DateTime64) exists since v22.8, I have opened issue #42615 for it.

**Comment by qoega** [comment]
This one also 02457_datediff_via_unix_epoch
https://s3.amazonaws.com/clickhouse-test-reports/0/5ec9b35bbbdad41f399979ae73d8fc135902fcb2/stateless_tests__msan__[1/3].html

**Comment by rvasin** [comment]
@qoega thank you for noticing it. The test 02457_datediff_via_unix_epoch fails on America/Hermosillo time-zone. I was able to re-produce it. I am working on it.

**Comment by rvasin** [comment]
I created a script to run 02457_datediff_via_unix_epoch test on all 594 time zones. The test fails only on 4 times zones:
America/Bahia_Banderas
America/Hermosillo
America/Mazatlan
Mexico/BajaSur

And the diff result is the same for all these 4 time zones:
```
11,12c11,12
< hour	263
< hour	263
---
> hour	264
> hour	264
14c14
< minute	-40
---
> minute	20
16c16
< second	-2400
---
> second	1200
```

**Comment by rvasin** [comment]
My guess is that for these 4 time zones there was a transition in 1970. And this transition is not correctly processed by CH. I already reported it [here](https://github.com/ClickHouse/ClickHouse/issues/38393#issuecomment-1184768923). So I will try to verify if transitions are correctly processed around 1970 for these 4 time zones.

**Comment by rvasin** [comment]
I have found that all these 4 time zones belong to Mexico. The list of world's time zones: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones

Each of 4 time zones redirects to one article: 
https://en.wikipedia.org/wiki/Time_in_Mexico
As the text in Wikipedia shows:

> For economic reasons, some metropolitan areas next to the U.S. border follow the [U.S. Daylight Saving Time schedule](https://en.wikipedia.org/wiki/Daylight_saving_time_in_the_United_States) instead of the Mexican schedule, resulting in those areas being an hour off from the rest of their respective states for a few weeks out of the year.

I think the bug in something related to DST time calculation/recording during DateLUT the generation in DateLUTImpl.cpp for these 4 time zones. I am working on it.

**Comment by rvasin** [comment]
I think the bug is in the following code in src/Common/DateLUTImpl.cpp:

```

cctz::time_zone::civil_lookup lookup = cctz_time_zone.lookup(date);

/// Ambiguity is possible if time was changed backwards at the midnight
/// or after midnight time has been changed back to midnight, for example one hour backwards at 01:00
/// or after midnight time has been changed to the previous day, for example two hours backwards at 01:00
/// Then midnight appears twice. Usually time change happens exactly at 00:00 or 01:00.

/// If transition did not involve previous day, we should use the first midnight as the start of the day,
/// otherwise it's better to use the second midnight.

std::chrono::time_point start_of_day_time_point = lookup.trans < lookup.post
    ? lookup.post /* Second midnight appears after transition, so there was a piece of previous day after transition */
    : lookup.pre;

start_of_day = std::chrono::system_clock::to_time_t(start_of_day_time_point);

```

**Comment by rvasin** [comment]
@rschu1ze I have found that for all 4 Mexican time zones above:
```
TZ=America/Hermosillo clickhouse-local --query "select toDateTime64('1970-01-01 00:00:00.000', 3);"
1969-12-31 23:00:00.000 
```
I created the issue #42653

Robert, right now 02457_datediff_via_unix_epoch is flaky because of these 4 time zones. 

To eliminate flaky behavior I suggest to create a mini PR to remove from the test all queries with usage of toDateTIme64 for hours, minutes and seconds. CH release 22.10 is approaching. What do you think?

**Comment by rschu1ze** [comment]
Sure, let's do that, thanks.

**Comment by rvasin** [comment]
@rschu1ze I created PR #42655 Please assign yourself as reviewer.
